## Load alignment of long reads from bam file generated by minimap2
# fbam <- readRDS('/project/gtplab/data_analysis/0_analyst/zhangzhe/Project/2024-03_PacBio_ONT_Illumina/R/file_bam_pacbio.rds');
# fbam <- fbam$`DG21062-p5748`["donor_vec"];
# fout <- "loaded/DG21062-p5748_donor.rds";
# what <- c("qname", "flag", "rname", "strand", "pos", "qwidth", "mapq", "cigar", "seq");
# tag <- c("NM", "AS", "SA", "tp", "cm", "nn", "s1", "s2", "ms", "rl", "de"); ## https://lh3.github.io/minimap2/minimap2.html
# unmapped <- FALSE;

LoadLongreadBam <- function(fbam, fout=NA, what=NA, tag=character(0), unmapped=FALSE) {
  ## fbam     Path and name of a bam file to be loaded
  ## fout     Write loaded data to an output file if given
  ## what     Which SAM fields to load; use ScanBamWhat() to use all default; 
  
  require(Rsamtools);
  
  ## Loading parameters
  if (identical(NA, what)) what <- scanBamWhat();
  param <- ScanBamParam(flag = scanBamFlag(isUnmappedQuery = unmapped), what=what, tag=tag);
  
  ## Loading bam
  aln <- scanBam(fbam, param = param)[[1]];
  
  ## Turn alignment to a data.frame
  tbl <- data.frame(aln[names(aln)!='tag']);
  
  ## Add extra tags to the table
  if (!is.null(aln$tag)) {
    xtr <- aln$tag;
    xtr <- xtr[!sapply(xtr, is.null)];
    if (length(xtr) > 0) tbl <- cbind(tbl, data.frame(xtr));
  }

  if (!is.na(fout)) saveRDS(tbl, fout); 
  
  invisible(tbl);
}