## Summarize alignment Bam file generated by minimap2
# faln  <- "/project/gtplab/data_analysis/0_analyst/zhangzhe/Project/2024-03_PacBio_ONT_Illumina/R/loaded/pacbio_MG21041DP-p6130_donor_vec.rds";
# aln   <- readRDS(faln);
# aln$rname <- 'p6130';

SummarizeMinimap2Alignment <- function(aln, ann = NA) {
  ## aln      A SAM table of with required columns
  ## ann      Annotated regions as a GRanges objects
  
  require(GenomicAlignments);
  
  aln <- aln[!is.na(aln$cigar), , drop=FALSE];

  
  ## Alignment ranges by CIGAR
  mrng0 <- cigarRangesAlongReferenceSpace(aln$cigar, pos=aln$pos, ops='M'); # Matching range in reference
  mrng1 <- cigarRangesAlongQuerySpace(aln$cigar, ops='M'); # Matching range in read
  
  drng0 <- cigarRangesAlongReferenceSpace(aln$cigar, pos=aln$pos, ops='D'); # Matching range in reference
  drng1 <- cigarRangesAlongQuerySpace(aln$cigar, ops='D'); # Matching range in read
  
  irng0 <- cigarRangesAlongReferenceSpace(aln$cigar, pos=aln$pos, ops='I'); # Matching range in reference
  irng1 <- cigarRangesAlongQuerySpace(aln$cigar, ops='I'); # Matching range in read
  
  ## Type of alignments (primary, secondary, supplementary)
  flg <- bamFlagAsBitMatrix(as.integer(aln$flag))[, c(9, 12)];
  flg <- cbind(primary=1-pmin(1, flg[, 1]+flg[, 2]), secondary=flg[, 1], supplementary=flg[, 2]);
  
  ## Read positions
  smm <- data.frame(read_id=as.vector(aln$qname), read_len=aln$qwidth);
  smm$read_start <- min(start(mrng1));
  smm$read_end   <- max(end(mrng1));
  
  ## Ref positions
  smm$ref <- aln$rname;
  smm$ref_start  <- min(start(mrng0));
  smm$ref_end    <- max(end(mrng0));
  smm$strand <- c('+'=1, '-'=-1, '*'=0)[as.vector(aln$strand)];
  
  ## Reverse positions in read if aligned to minus strand
  pos1 <- smm$read_start;
  pos2 <- smm$read_end;
  pos1[smm$strand==-1] <- smm$read_len[smm$strand==-1] - pos1[smm$strand==-1] + 1;
  pos2[smm$strand==-1] <- smm$read_len[smm$strand==-1] - pos2[smm$strand==-1] + 1;
  smm$read_start <- pmin(pos1, pos2);
  smm$read_end   <- pmax(pos1, pos2);
  
  ## Other alignment info
  smm <- cbind(smm, flg);
  smm$flag <- aln$flag;
  smm$mapq <- aln$mapq;
  
  ## Total number of matching (including mismatch), deleted and inserted bases. 
  smm$nmatch     <- sum(width(mrng0));
  smm$ndeletion  <- sum(width(drng0));
  smm$ninsertion <- sum(width(irng1));
  smm$max_deletion  <- pmax(0, max(width(drng0)));
  smm$max_insertion <- pmax(0, max(width(irng1)));
  
  smm0 <- smm[smm$primary==1, , drop=FALSE];
  smm1 <- smm[smm$supplementary==1, , drop=FALSE];
  
}