## Summarize alignment Bam file generated by minimap2
# fbam <- readRDS('/project/gtplab/data_analysis/0_analyst/zhangzhe/Project/2024-03_PacBio_ONT_Illumina/R/file_bam_pacbio.rds');
# fbam <- fbam$`DG21062-p5748`["donor_vec"];
# faln  <- "/project/gtplab/data_analysis/0_analyst/zhangzhe/Project/2024-03_PacBio_ONT_Illumina/R/loaded/pacbio_MG21041DP-p6130_donor_vec.rds";
# aln   <- readRDS(faln);
# aln$rname <- 'p6130';

# Load a minimap2 generated bam file of long reads, using default setting
# Summarize alignment into a text table
LoadSummarizeMinimap2Alignment <- function(fbam, fout=NA, ann=NA) {
  ## fbam     A bam file generated by minimap2 alignment of long reads
  ## fout     Save loaded data to a output file if specified
  ## ann      Annotated regions as a GRanges objects
  
  require(Rsamtools); 
  
  ## Default SAM fields and tags to load
  what <- c("qname", "flag", "rname", "strand", "pos", "qwidth", "mapq", "cigar", "seq");
  tag <- c("NM", "AS", "SA", "tp", "cm", "nn", "s1", "s2", "ms", "rl", "de"); ## https://lh3.github.io/minimap2/minimap2.html
  unmapped <- FALSE;
  
  ## Loading parameters
  if (identical(NA, what)) what <- scanBamWhat();
  param <- ScanBamParam(flag = scanBamFlag(isUnmappedQuery = unmapped), what=what, tag=tag);
  
  ## Loading bam
  aln <- scanBam(fbam, param = param)[[1]];
  
  ## Turn alignment to a data.frame
  tbl <- data.frame(aln[names(aln)!='tag']);
  
  ## Add extra tags to the table
  if (!is.null(aln$tag)) {
    xtr <- aln$tag;
    xtr <- xtr[!sapply(xtr, is.null)];
    if (length(xtr) > 0) tbl <- cbind(tbl, data.frame(xtr));
  }
  
  ## Write loaded alignment to a file
  if (!is.na(fout)) saveRDS(tbl, fout); 
  
  smm <- SummarizeMinimap2Alignment(tbl, ann);
  
  invisible(smm);
}

SummarizeMinimap2Alignment <- function(aln, ann = NA, all.indel = TRUE) {
  ## aln        A SAM table of with required columns
  ## ann        Annotated regions as a GRanges objects
  ## all.inde   Whether to include position/length of all INDELs in the result table
  
  require(GenomicAlignments);
  
  aln <- aln[!is.na(aln$cigar), , drop=FALSE];
  
  ## Alignment ranges by CIGAR
  mrng0 <- cigarRangesAlongReferenceSpace(aln$cigar, pos=aln$pos, ops='M'); # Matching range in reference
  mrng1 <- cigarRangesAlongQuerySpace(aln$cigar, ops='M'); # Matching range in read
  
  drng0 <- cigarRangesAlongReferenceSpace(aln$cigar, pos=aln$pos, ops='D'); # Matching range in reference
  drng1 <- cigarRangesAlongQuerySpace(aln$cigar, ops='D'); # Matching range in read
  
  irng0 <- cigarRangesAlongReferenceSpace(aln$cigar, pos=aln$pos, ops='I'); # Matching range in reference
  irng1 <- cigarRangesAlongQuerySpace(aln$cigar, ops='I'); # Matching range in read
  
  ## Type of alignments (primary, secondary, supplementary)
  flg <- bamFlagAsBitMatrix(as.integer(aln$flag))[, c(9, 12)];
  flg <- cbind(primary=1-pmin(1, flg[, 1]+flg[, 2]), secondary=flg[, 1], supplementary=flg[, 2]);
  
  ## Read positions
  smm <- data.frame(read_id=as.vector(aln$qname), read_len=aln$qwidth);
  smm$read_start <- min(start(mrng1));
  smm$read_end   <- max(end(mrng1));
  
  ## Ref positions
  smm$ref <- aln$rname;
  smm$ref_start  <- min(start(mrng0));
  smm$ref_end    <- max(end(mrng0));
  smm$strand <- c('+'=1, '-'=-1, '*'=0)[as.vector(aln$strand)];
  
  ## Reverse positions in read if aligned to minus strand
  pos1 <- smm$read_start;
  pos2 <- smm$read_end;
  pos1[smm$strand==-1] <- smm$read_len[smm$strand==-1] - pos1[smm$strand==-1] + 1;
  pos2[smm$strand==-1] <- smm$read_len[smm$strand==-1] - pos2[smm$strand==-1] + 1;
  smm$read_start <- pmin(pos1, pos2);
  smm$read_end   <- pmax(pos1, pos2);
  
  ## Other alignment info
  smm <- cbind(smm, flg);
  smm$flag <- aln$flag;
  smm$mapq <- aln$mapq;
  
  ## Total number of matching (including mismatch), deleted and inserted bases. 
  smm$nmatch     <- sum(width(mrng0));
  smm$ndeletion  <- sum(width(drng0));
  smm$ninsertion <- sum(width(irng1));
  smm$max_deletion  <- pmax(0, max(width(drng0)));
  smm$max_insertion <- pmax(0, max(width(irng1)));
  
  if (all.indel) {
    dpos <- paste(start(drng0), collapse=';'); # Position of deletion is the position of the first deleted base
    dlen <- paste(width(drng0), collapse=';'); 
    
    ipos <- paste(end(irng0), collapse=';'); # Postion of the insertion is the position of the base before the insertion
    ilen <- paste(width(irng1), collapse=';');
    
    smm$pos_deletion  <- dpos;
    smm$pos_insertion <- ipos;
    
    smm$len_deletion  <- dlen;
    smm$len_insertion <- ilen;
  }
  
  invisible(smm); 
}
